diff --git a/src/tplink-safeloader.c b/src/tplink-safeloader.c
index b7c73c4..f11caeb 100644
--- a/src/tplink-safeloader.c
+++ b/src/tplink-safeloader.c
@@ -1807,6 +1807,63 @@ static struct device_info boards[] = {
 		.last_sysupgrade_partition = "file-system",
 	},
 
+	/** Firmware layout for the Deco M4R v3 */
+	{
+		.id = "DECO-M4R-V3",
+		.vendor = "",
+		.support_list =
+			"SupportList:\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:00000000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:55530000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:45550000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:43410000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:4A500000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:41550000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:54570000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:42340000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:49440000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:5A470000}\n"
+			"{product_name:M4R,product_ver:3.0.0,special_id:4B520000}\n",
+		.part_trail = 0x00,
+		.soft_ver = SOFT_VER_DEFAULT,
+
+		.partitions = {
+			{"SBL1", 0x00000, 0x30000},
+			{"boot-config_0", 0x30000, 0x10000},
+			{"MIBIB", 0x40000, 0x10000},
+			{"boot-config_1", 0x50000, 0x10000},
+			{"QSEE", 0x60000, 0x60000},
+			{"CDT", 0xc0000, 0x10000},
+			{"DDRPARAMS", 0xd0000, 0x10000},
+			{"uboot-env", 0xe0000, 0x10000},
+			{"fs-uboot@0", 0xf0000, 0x80000},
+			{"radio", 0x170000, 0x0fff0},
+			{"bluetooth-XTAL", 0x17fff0, 0x00010},
+			{"default-mac", 0x180000, 0x01000},
+			{"device-id", 0x182000, 0x01000},
+			{"product-info", 0x183000, 0x05000},
+			{"support-list", 0x190000, 0x10000},
+			{"user-config", 0x200000, 0x10000},
+			{"device-config", 0x210000, 0x10000},
+			{"group-info", 0x220000, 0x10000},
+			{"partition-table@0", 0x230000, 0x02000},
+			{"os-image@0", 0x240000, 0x320000},
+			{"file-system@0", 0x560000, 0xa40000},
+			{"fs-uboot@1", 0xfa0000, 0x80000},
+			{"os-image@1", 0x1020000, 0x400000},
+			{"file-system@1", 0x1420000, 0xbe0000},
+			{NULL, 0, 0}
+		},
+
+		.partition_names.partition_table = "partition-table@0",
+		.partition_names.soft_ver = NULL,
+		.partition_names.os_image = "os-image@1",
+		.partition_names.file_system = "file-system@1",
+
+		.first_sysupgrade_partition = "os-image@1",
+		.last_sysupgrade_partition = "file-system@1",
+	},
+
 	/** Firmware layout for the Deco M4R v4 */
 	{
 		.id     = "DECO-M4R-V4",
